### ISSUE TITLE
User Session Count Limiter not working for some users

### ISSUE THREAD

**Issue by nathangerday** [keycloak/keycloak#14689]
### Describe the bug

Hi, 
We recently implemented a new Authentication flow with an "User Session Count Limiter" step to put a hard limit for each user in the realm at 50 sessions. 
This was tested with an account and seemed to work properly, each session after the 50th was terminating the oldest without any issue. 
However, now that we've deployed this flow in our production server, we noticed that some users still manage to go above this limit. We tried to logout all their sessions but they still manage to get to way more than 50 sessions ( above 1000 for some ). 
The limit was added on the Direct Grant Flow and the Browser Flow.
And it doesn't seem that they use another flow to bypass this, since we are getting the following logs : `[org.keycloak.authentication.authenticators.sessionlimits.UserSessionLimitsAuthenticator] (executor-thread-6727) Too many session in this realm for the current user. Session count: 299`

This feels like a bug to me but if we are missing anything or can provide any more information, please tell us so.
Thanks in advance !

### Version

18.0.0

### Expected behavior

No users should have a session count above 50.

### Actual behavior

Many users are getting sessions count above 50 (up to 1200+) 

### How to Reproduce?

We haven't been able to reproduce this with our own accounts.

**Comment by lexcao** [comment]
> Note that the user session limits should be added to your bound Browser flow, Direct grant flow, Reset credentials and also to any Post broker login flow.

As the [docs](https://www.keycloak.org/docs/latest/server_admin/index.html#script-authenticator) said, are there any other flows for users to log in to your system?
If that,  Reset credentials and Post broker login flow should also configure the Session Limit.

Could you please check with the logs to see if the login method of the user who exceeded the 1000+ limit of sessions?

**Comment by mposolda** [comment]
@douglaspalmer I think that it is missed the configuration of some other flow as mentioned by @lexcao , but it is possible that our docs should be improved.  Assigning to you to improve the docs and make it slightly more clear.

For example:
- We can change existing docs to contain WARNING / NOTE (in the box) to highlight this that reset-password, direct-grant and IDP login require further setup
- We should explicitly mention that IDP brokering requires configuration of "Post Broker Login" flow in orther that session limits is applied. (Currently there is only note "also to any Post Broker Login Flow", but it is not mentioned anywhere that IDP brokering requires this setup and people using IDP may not necessarily know what "Post Broker Login Flow" is).

**Comment by drohwer89** [comment]
We have noticed a similar behavior in our test environment. In our test environment, we are running 2 Keycloak instances in a GKE cluster and have configured the session cache to be replicated. We are using the uptime check provided by GCP. This uptime check creates about 6 sessions per minute. We provide a user and a client that are used only for this purpose. For the Direct Grant flow, we have set a limit of 10 sessions per user and client. This is the only flow that will be used by the uptime check user. Initially, there are only 10 sessions for this user through the uptime check client. But after 15 or 20 minutes, there were already 11 sessions for this user via the uptime check client. The number of sessions increased slowly at first, but afer some time it exploded. After 12 hours, we finally had 3500 sessions for the uptime check user. This results in slow session creation (over 2500 ms) for the uptime check user, while other users can still create sessions quickly (around 120 ms)

We did not understand this behavior and took a closer look at the UserSessionLimitsAuthenticator. We noticed that the logoutOldestSession method only terminates the oldest session:
```   
private void logoutOldestSession(List<UserSessionModel> userSessions) {
        logger.info("Logging out oldest session");
        Optional<UserSessionModel> oldest = userSessions.stream().sorted(Comparator.comparingInt(UserSessionModel::getLastSessionRefresh)).findFirst();
        oldest.ifPresent(userSession -> AuthenticationManager.backchannelLogout(session, userSession, true));
    }
```


We think that our setup, combined with the number of sessions created by the uptime check, dramatically exacerbates the problem. However, we do not believe that the current implementation will result in the desired behavior. We believe that any session that exceeds the configured limit should be terminated. Therefore, I have changed the implementation so that the oldest x sessions are terminated. 

The behavior in our test environment may be due to our setup. However, there is another use case that does not work with the current implementation. If the session limit per client is set to 10 and you want to lower it to 5, this change will have no effect. There will still be 10 sessions per client because only the oldest session will be terminated if the number of sessions is above the limit. I adjusted the testClientSessionCountExceededAndOldestSessionRemovedDirectGrantFlow test to show that the optimized implementation behaves as expected in this use case.
